lib/src/component2_suggestors/setstate_updater.dart

```diff
@@ -0,0 +1,45 @@
+// Copyright 2019 Workiva Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import 'package:analyzer/analyzer.dart';
+import 'package:codemod/codemod.dart';
+
+import '../constants.dart';
+import '../util.dart';
+
+/// Suggestor that updates a [setState] call to [setStateWithUpdater] in the
+/// case that the first argument is a function.
+class SetStateUpdater extends GeneralizingAstVisitor
+    with AstVisitingSuggestorMixin
+    implements Suggestor {
+  @override
+  visitMethodInvocation(MethodInvocation node) {
+    super.visitMethodInvocation(node);
+
+    if (node.argumentList.arguments.isEmpty) return;
+
+    final firstArg = node.argumentList.arguments.first;
+
+    if (node.methodName.name == 'setState') {
+      if (firstArg is MethodInvocation &&
+          firstArg.methodName.name == 'newState') return;
+
+      if (firstArg is MethodInvocation || firstArg is FunctionExpression) {
+        int length = node.toString().indexOf('(');
+
+        yieldPatch(node.offset, node.offset + length, 'setStateWithUpdater');
```

Comment:
#nit maybe you could change this to use `methodName` so then you won't have to find `length`
```suggestion
        yieldPatch(node.methodName.offset, node.methodName.end, 'setStateWithUpdater');
```

Reply:
That's a really good idea. In my original attempt when I only looking for a `MethodInvocation`, that would have been ideal. Now that a `FunctionExpression` is involved though, it's not guaranteed that `methodName` will exist.. If there's a real detriment to using `length` as it is, definitely let me know and that would make it worth refactoring. Otherwise (after experimenting with refactoring), it seems simpler to keep it as is. That was a suuuper good catch though 😄

Reply:
Oh that's true! I didn't think about it being a `FunctionExpression`.

Reply:
Update: Sydney was 100% correct and I was looking at it incorrectly. That has been updated!